app/models/manageiq/providers/hawkular/middleware_manager/middleware_diagnostic_report.rb

```diff
@@ -0,0 +1,139 @@
+
+      self.status = STATUS_RUNNING
+      save!
+
+      callback = proc do |on|
+        on.success(&method(:jdr_report_succeded))
+        on.failure(&method(:jdr_report_failed))
+      end
+
+      @connection = ext_management_system.connect.operations(true)
+      @finish_signal = Queue.new
+      @connection.export_jdr(middleware_server.ems_ref, true, &callback)
+
+      Timeout.timeout(::Settings.ems.ems_hawkular.jdr.generation_timeout.to_i_with_method) { @finish_signal.deq }
+    rescue Timeout::Error
+      self.status = STATUS_ERROR
+      self.error_message = _('Reached generation timeout.')
+      save!
+    rescue => ex
+      self.status = STATUS_ERROR
+      self.error_message = ex.to_s
+      save!
+    ensure
+      @connection.close_connection! if @connection
+    end
+
+    private
+
+    def set_queued_date
+      self.queued_at = Time.current if queued?
+    end
+
+    def enqueue_job
+      unless queued?
+        $mw_log.debug("#{log_prefix} JDR report registry [#{id}] for server [#{middleware_server.ems_ref}] created with status [#{status}].")
+        return
+      end
+
+      job = MiqQueue.submit_job(
+        :class_name  => self.class.name,
+        :instance_id => id,
+        :role        => 'ems_operations',
+        :method_name => 'generate_diagnostic_report',
+        :msg_timeout => ::Settings.ems.ems_hawkular.jdr.generation_timeout.to_i_with_method + 30.seconds
+      )
+
+      EmsEvent.add_queue(
+        'add', ems_id,
+        :ems_id          => ems_id,
+        :source          => 'EVM',
+        :timestamp       => Time.zone.now,
+        :event_type      => 'hawkular_event',
+        :message         => _('Generation of JDR report was requested by a user.'),
+        :middleware_ref  => middleware_server.ems_ref,
+        :middleware_type => 'MiddlewareServer',
+        :username        => requesting_user
+      )
+
+      $mw_log.info("#{log_prefix} JDR report [#{id}] for server [#{middleware_server.ems_ref}] enqueued with job #{job.id}.")
+    end
+
+    def jdr_report_succeded(data)
+      reload
+      self.class.transaction do
+        if binary_blob
+          $mw_log.debug("#{log_prefix} JDR report [#{id}] [#{binary_blob.name}] will be overwritten.")
+          binary_blob.name = data['fileName']
+          binary_blob.data_type = 'zip'
+        else
+          self.binary_blob = BinaryBlob.create(:name => data['fileName'], :data_type => 'zip')
+        end
+
+        binary_blob.binary = data[:attachments]
+        self.status = STATUS_READY
+        save!
+      end
+
+      $mw_log.info("#{log_prefix} Generation of JDR report [#{id}] [#{binary_blob.name}] succeded.")
+      @finish_signal << STATUS_READY
+    end
+
+    def jdr_report_failed(error)
+      $mw_log.warn("#{log_prefix} Generation of JDR report [#{id}] failed: #{error}.")
+      self.status = STATUS_ERROR
+      self.error_message = error
+      save!
+
+      @finish_signal << STATUS_READY
```

Comment:
The queue `@finish_signal` is for blocking the main thread when the request to generate JDR is sent to Hawkular-services command gateway [right here](https://github.com/ManageIQ/manageiq-providers-hawkular/pull/30/files#diff-c6947bd67b4592e6ae39ce22e286f625R58). You know: because websockets are being used for cmd-gw operations, `export_jdr` method is asynchronous. The callbacks are ran in the context of a secondary thread and in order to unblock the main thread I just pushed something to `@finish_signal`. It could have been any arbitrary value, it doesn't mean anyting nor it's used. The only purpose is to unblock the main thread once JDR fails or succeeds to close the websocket connection [here](https://github.com/ManageIQ/manageiq-providers-hawkular/pull/30/files#diff-c6947bd67b4592e6ae39ce22e286f625R68) and ensure that the secondary thread is terminated.

Should the value cause confusion, I could replace it with an arbitrary `string`, a `:symbol` or a boolean value.